add summary of JupyterHub MockHub class and App fixture (#12)
jupyterhub/jupyterhub-pytest-plugin: docs/source/fixtures.md

# Fixtures and Mocks of the `jupyterhub` package

(jupyterhub-mockhub)=
## [The MockHub Class](https://github.com/jupyterhub/jupyterhub/blob/e4f72c9eeb4cd308ff5cbcf21142b2cb0a0345e4/jupyterhub/tests/mocking.py#L220) 

> It is a subclass of `JupyterHub` that allows for easier testing of the `JupyterHub` class and other components that require a running instance of JupyterHub. It does this by providing [mock](https://docs.python.org/3/library/unittest.mock.html) implementations of particular functionality of a JupyterHub instance. It also allows for easier setup and cleanup of the testing environment.

Reviewer: "implementations of particular functionality"
What does "particular functionality" mean?

Author: > "implementations of particular functionality" What does "particular functionality" mean?

Hey @manics,
By "particular functionality", I mean that it provides some specific features belonging to the JupyterHub class, sometimes modified for testing.
Is there a way you think I could word it better?

Reviewer 2: @Sheila-nk, I believe writing the explanation you just gave 👍🏼 , together with an example of these pieces that are being simplified by mocking (like initializing the hub database, initializing the hub and stopping and cleaning up the resources used by this MockHub during a test) should be enough for a more clear picture of the things that are mocked. What do you think?

Author: Yes, I think that works. I will add that information to give a clearer picture.

For more information, check out this [pytest documentation](https://docs.pytest.org/en/latest/explanation/fixtures.html) on fixtures.

A **Mock** is an object that simulates the behavior of another object such as a class or function. They are used to simulate the behavior of real objects for testing purposes.
For more information on Mocks, check out this [unittest documentation](https://docs.python.org/3/library/unittest.mock.html) on the mock module.
```{admonition} Example
The [init_db](https://github.com/jupyterhub/jupyterhub/blob/336d7cfcfaf74087e4ee467d5e3d3bec0c25c3d0/jupyterhub/app.py#L1804) function in JupyterHub's `app` module initializes a connection to a database using SQLAlchemy's ORM (Object-Relational Mapper).
However, the mock [init_db](https://github.com/jupyterhub/jupyterhub/blob/336d7cfcfaf74087e4ee467d5e3d3bec0c25c3d0/jupyterhub/tests/mocking.py#L295) function in JupyterHub's `mocking` module initializes a database connection for the mocked JupyterHub application instance by calling the `init_db` function of the JupyterHub superclass but also has a `test_clean_db` attribute to ensure that the database is reset to a clean state before running tests.
```

> ## [The MockHub Class](https://github.com/jupyterhub/jupyterhub/blob/e4f72c9eeb4cd308ff5cbcf21142b2cb0a0345e4/jupyterhub/tests/mocking.py#L220) 

Reviewer 2: So we can cross-reference it from the App fixture section:

```suggestion
(jupyterhub-mockhub)=
## [The MockHub Class](https://github.com/jupyterhub/jupyterhub/blob/e4f72c9eeb4cd308ff5cbcf21142b2cb0a0345e4/jupyterhub/tests/mocking.py#L220) 
```

> It is a subclass of `JupyterHub` that allows for easier testing of the `JupyterHub` class and other components that require a running instance of JupyterHub. It does this by providing mock implementations of particular functionality of a JupyterHub instance. It also allows for easier setup and cleanup of the testing environment.

Reviewer 2: ```suggestion
It is a subclass of `JupyterHub` that allows for easier testing of the `JupyterHub` class and other components that require a running instance of JupyterHub. It does this by providing [mock implementations](https://docs.python.org/3/library/unittest.mock.html) of particular functionality of a JupyterHub instance. It also allows for easier setup and cleanup of the testing environment.
```

```{admonition} Example
The `conftest` module of the `DockerSpawner` repository imports and uses the JupyterHub `MockHub` class. This can be seen [here](https://github.com/jupyterhub/dockerspawner/blob/8503af69161a3a543cc613f93ce7951ad30a1912/tests/conftest.py#L26).

It sets the `hub_ip` and `hub_connect_ip` attributes of the mocked instance.
These attributes specify the IP addresses that the mock instance of JupyterHub will be listening in on and connecting to.
```

> ## [The App Fixture](https://github.com/jupyterhub/jupyterhub/blob/e4f72c9eeb4cd308ff5cbcf21142b2cb0a0345e4/jupyterhub/tests/conftest.py#L61)
> It is used to provide a mock `JupyterHub` application instance for testing. It is also responsible for starting the mock `JupyterHub` application instance and stopping it after all tests in the module have been completed.

Reviewer 2: Let's:
- add a cross reference to this section in case we will need it later
- reword a bit to avoid confusion with the MockHub class (a MockHub is a JupyterHub object with some of the functions mocked and an `app` is a "running", initialized MockHub, i.e. a lightweight running JupyterHub)
- add a link to the `fixture` docs as mentioned in a comment above
https://myst-parser.readthedocs.io/en/latest/intro.html#cross-referencing

```suggestion
(jupyterhub-app-fixture)=
## [The App Fixture](https://github.com/jupyterhub/jupyterhub/blob/e4f72c9eeb4cd308ff5cbcf21142b2cb0a0345e4/jupyterhub/tests/conftest.py#L61)
The `app` [fixture](https://docs.pytest.org/en/latest/explanation/fixtures.html) is used to provide a lightweight, running`JupyterHub` application for testing. It is responsible for starting the mock `JupyterHub` application instance, the [MockHub](mockhub) and stopping it after all tests in the module have been completed.
```

# Fixtures and Mocks
A **Fixture** is a function that is used to prepare and clean up the environment for a test function. Fixtures can be used to set up test data, test environment, and other resources that are needed by test functions.
For more information, check out this [pytest documentation](https://docs.pytest.org/en/latest/explanation/fixtures.html) on fixtures.

A **Mock** is an object that simulates the behavior of another object such as a class or function. They are used to simulate the behavior of real objects for testing purposes.
For more information on Mocks, check out this [unittest documentation](https://docs.python.org/3/library/unittest.mock.html) on the mock module.

*Example:*

The [init_db](https://github.com/jupyterhub/jupyterhub/blob/336d7cfcfaf74087e4ee467d5e3d3bec0c25c3d0/jupyterhub/app.py#L1804) function in JupyterHub's `app` module initializes a connection to a database using SQLAlchemy's ORM (Object-Relational Mapper).

> However, the mock [init_db](https://github.com/jupyterhub/jupyterhub/blob/336d7cfcfaf74087e4ee467d5e3d3bec0c25c3d0/jupyterhub/tests/mocking.py#L295) function in JupyterHub's `mocking` module initializes a database connection for the mocked JupyterHub application instance by calling the `init_db` function of the JupyterHub superclass but also has a `test_clean_db` attribute to ensure that the database is reset to a clean state before running tests.

Author: Sorry, @GeorgianaElena, I don't completely understand this change, please clarify. You are suggesting I add the link to the docs on mocks and fixtures(currently fixtures.md) in the README and remove that part completely from the README?

> However, the mock [init_db](https://github.com/jupyterhub/jupyterhub/blob/336d7cfcfaf74087e4ee467d5e3d3bec0c25c3d0/jupyterhub/tests/mocking.py#L295) function in JupyterHub's `mocking` module initializes a database connection for the mocked JupyterHub application instance by calling the `init_db` function of the JupyterHub superclass but also has a `test_clean_db` attribute to ensure that the database is reset to a clean state before running tests.

Reviewer 2: @Sheila-nk, I was suggesting to remove the dupe from `fixtures.md` and then the first time the word `mock` shows up, add a link to https://docs.python.org/3/library/unittest.mock.html and same for `fixture`, add a link to https://docs.pytest.org/en/latest/explanation/fixtures.html. I will add some inline suggestions below also.

This way, people coming to the docs without first reading the readme, will have a handy reference of what a mock & fixture is.

> However, the mock [init_db](https://github.com/jupyterhub/jupyterhub/blob/336d7cfcfaf74087e4ee467d5e3d3bec0c25c3d0/jupyterhub/tests/mocking.py#L295) function in JupyterHub's `mocking` module initializes a database connection for the mocked JupyterHub application instance by calling the `init_db` function of the JupyterHub superclass but also has a `test_clean_db` attribute to ensure that the database is reset to a clean state before running tests.

Author: Oh great! I understand now. Thank you

**[JupyterHub MockHub Class](https://github.com/jupyterhub/jupyterhub/blob/e4f72c9eeb4cd308ff5cbcf21142b2cb0a0345e4/jupyterhub/tests/mocking.py#L220)**: 

> It is a subclass of `JupyterHub` that allows for easier testing of the `JupyterHub` class and other components that require a running instance of JupyterHub. It does this by providing mock implementations of particular functionality of a JupyterHub instance. It also allows for easier setup and cleanup of the testing environment.

Reviewer 2: Because it is not yet properly documented in the template repository unfortunately :(, I'm not sure you noticed  that the docs setup enables an extension called [myst-parser](https://myst-parser.readthedocs.io/en/latest):

https://github.com/jupyterhub/jupyterhub-pytest-plugin/blob/45d3b2b76c9afef533fa6fc494ffd9fb56da3959/docs/source/conf.py#L41

This extension enables us with multiple option when writing Markdown documentation. This piece of documentation describes its core syntax https://myst-parser.readthedocs.io/en/latest/syntax/syntax.html.

Using this syntax, instead of emphasizing sections with only bold text, I would use headings and maybe a custom [`Example` directive](https://myst-parser.readthedocs.io/en/latest/syntax/roles-and-directives.html#directives-a-block-level-extension-point) instead, in order to make them easier to read. What do you think?

Author: I have just tried this out and it makes it all easier to read. I like it, thank you 😄
